# Are there drawbacks to using the `Decimal` class?

- Yes!

**Drawbacks**

1. Need to construct the objects from strings or tuples
    - Therefore, not as easy to create
2. Not all mathematical functions are defined explicitly for `Decimal` objects
    - So we're forced to lose precision by using the `math` module
3. Uses more memory than floats
4. Slower performance than floats

# Examples

In [1]:
from decimal import Decimal
import sys

**Memory**

In [2]:
a = 3.1415
b = Decimal('3.1415')
sys.getsizeof(a), sys.getsizeof(b)

(24, 104)

- As we can see, `a` takes up 24 bytes while `b` takes up 104 bytes
    - More than 4 times the size

**Performance**

In [3]:
import time

In [4]:
def run_float(n=1):
    for i in range(n):
        a = 3.1415

In [5]:
def run_decimal(n=1):
    for i in range(n):
        b = Decimal('3.1415')

In [6]:
n = 10000000

In [7]:
start = time.perf_counter()
run_float(n)
end = time.perf_counter()
total = end - start
total

0.2365252

In [8]:
start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
total = end - start
total

1.957491899999999

- As we can see, it took almost 8 times as long to run the `Decimal` version

In [10]:
def run_float_add(n=1):
    a = 3.1415
    for i in range(n):
        a + a

In [11]:
def run_decimal_add(n=1):
    b = Decimal('3.1415')
    for i in range(n):
        b + b

In [12]:
start = time.perf_counter()
run_float_add(n)
end = time.perf_counter()
total = end - start
total

0.369741399999981

In [13]:
start = time.perf_counter()
run_decimal_add(n)
end = time.perf_counter()
total = end - start
total

0.7276353000000029

- Took twice as long

**Mathematical functions**

In [14]:
import math

In [15]:
n = 5000000

In [16]:
def run_float_sqrt(n=1):
    a = 3.1415
    for i in range(n):
        math.sqrt(a)

In [17]:
def run_decimal_sqrt(n=1):
    b = Decimal('3.1415')
    for i in range(n):
        b.sqrt()

In [18]:
start = time.perf_counter()
run_float_sqrt(n)
end = time.perf_counter()
total = end - start
total

0.639400500000022

In [19]:
start = time.perf_counter()
run_decimal_sqrt(n)
end = time.perf_counter()
total = end - start
total

13.408491799999979

- For this function, more than 20x slower!

- In conclusion, **we want to use `Decimal` objects sparingly**